GVFS/GVFS.Common/NuGetUpgrader/NuGetFeed.cs

```diff
@@ -0,0 +1,226 @@
+        /// </summary>
+        /// <param name="packageId"></param>
+        /// <returns>List of packages that match query parameters</returns>
+        public virtual async Task<IList<IPackageSearchMetadata>> QueryFeedAsync(string packageId)
+        {
+            try
+            {
+                PackageMetadataResource packageMetadataResource = await this.sourceRepository.GetResourceAsync<PackageMetadataResource>();
+                IEnumerable<IPackageSearchMetadata> queryResults = await packageMetadataResource.GetMetadataAsync(
+                    packageId,
+                    includePrerelease: true,
+                    includeUnlisted: true,
+                    sourceCacheContext: this.sourceCacheContext,
+                    log: this.nuGetLogger,
+                    token: CancellationToken.None);
+
+                // TODO: consider working with just the IEnumerable
+                return queryResults.ToList();
+            }
+            catch (Exception ex)
+            {
+                // TODO: consider utility method to help trace exceptions (possibly including stack infomration);
+                throw new Exception($"Failed to query the NuGet package feed due to error: {ex.Message}", ex);
+            }
+        }
+
+        /// <summary>
+        /// Download the specified packageId from the NuGet feed.
+        /// </summary>
+        /// <param name="packageId">PackageIdentity to download.</param>
+        /// <returns>Path to the downloaded package.</returns>
+        public virtual async Task<string> DownloadPackage(PackageIdentity packageId)
+        {
+            string downloadPath = Path.Combine(this.downloadFolder, $"{this.feedName}.zip");
+            PackageDownloadContext packageDownloadContext = new PackageDownloadContext(
+                this.sourceCacheContext,
+                this.downloadFolder,
+                true);
+
+            DownloadResource downloadResource = await this.sourceRepository.GetResourceAsync<DownloadResource>();
+
+            using (DownloadResourceResult downloadResourceResult = await downloadResource.GetDownloadResourceResultAsync(
+                       packageId,
+                       packageDownloadContext,
+                       globalPackagesFolder: string.Empty,
+                       logger : this.nuGetLogger,
+                       token: CancellationToken.None))
+            {
+                // Check download result status
+                using (FileStream fileStream = File.Create(downloadPath))
+                {
+                    downloadResourceResult.PackageStream.CopyTo(fileStream);
+                }
+            }
+
+            return downloadPath;
+        }
+
+        /// <summary>
+        /// Implementation of logger used by NuGet library. It takes all output
+        /// and redirects it to the GVFS logger.
+        /// </summary>
+        public class Logger : ILogger
```

Comment:
Should `Logger` class be moved out into GVFS.Common? If this is going to be there only temporarily during testing, then its ok.

Reply:
This is the logger for NuGet (and is a nested class inside of NuGetFeed) - it has to be accessible for interacting with the NuGet APIs, so I kept it with this class.